### PR TITLE
Infer better variable name when initializing from an anonymous type

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/21602

### CODE REVIEW

File: src/Features/CSharpTest/IntroduceVariable/IntroduceVariableTests.cs [hunk 1]
```diff
@@ -8984,4 +8984,33 @@ public class C
...
+            """,
+            """
+            using System;
+
+            class C
+            {
+                void M()
+                {
+                    string {|Rename:y|} = DateTime.Now.ToString();
+                    var x = new { y = y };
```
Comment (ToddGrun) [inline src/Features/CSharpTest/IntroduceVariable/IntroduceVariableTests.cs:9011]: > var x = new { y = y }; [](http://example.com/codeflow?start=19&length=23)

doesn't match desired output from bug, is that a separate concern?

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SemanticModelExtensions.cs [hunk 2]
```diff
@@ -362,44 +362,48 @@ public static string GenerateNameForExpression(
...
                 var word = foreachStatement.Identifier.ValueText.ToCamelCase();
                 return CodeAnalysis.Shared.Extensions.SemanticModelExtensions.Pluralize(word);
             }
-            else
+
+            if (current.Parent is AnonymousObjectMemberDeclaratorSyntax { NameEquals: { } nameEquals } anonymousObjectMemberDeclarator &&
+                anonymousObjectMemberDeclarator.Expression == current)
             {
-                break;
+                return nameEquals.Name.Identifier.ValueText.ToCamelCase();
```
Comment (ToddGrun) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SemanticModelExtensions.cs:391]: >nameEquals.Name.Identifier.ValueText.ToCamelCase(); [](http://example.com/codeflow?start=23&length=51)

There isn't an option for determining how to case this?

File: src/Features/CSharpTest/IntroduceVariable/IntroduceVariableTests.cs (hunk 1 above)
Comment (ToddGrun) [reply src/Features/CSharpTest/IntroduceVariable/IntroduceVariableTests.cs:9011]: chatted offline

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SemanticModelExtensions.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SemanticModelExtensions.cs:391]: Probably.  But this just follows what we already do in this method.  And it doesn't look like we have customer complaining about locals naming (looks like it's pretty universal).

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**